src/leo_s3_user.erl

```
@@ -70,8 +72,72 @@ create_table(Mode, Nodes) ->
+                                      HashType::bcrypt | md5 | md5_with_salt,
+                                      Cause::any()).
+create(UserId, Password, HashType) ->
+    case find_by_id(UserId) of
+        not_found ->
+            Fun = fun() ->
+                          create_fun(UserId, Password, HashType)
+                  end,
+            case catch mnesia:activity(sync_transaction, Fun) of
+                {ok,_} = Ret ->
+                    Ret;
+                {_, Cause} ->
+                    {error, Cause}
+            end;
+        {ok,_} ->
+            {error, already_exists};
+        {error, Cause} ->
+            {error, Cause}
+    end.
+
+
+%% @private
+create_fun(UserId, Password, HashType) ->
+    CreatedAt = leo_date:now(),
+    Digest = hash_and_salt_password(HashType, Password, CreatedAt),
+
+    %% insert a user-info
+    case mnesia:write(?USERS_TABLE,
+                      #?S3_USER{id = UserId,
+                                password = Digest,
+                                created_at = CreatedAt,
+                                updated_at = CreatedAt}, write) of
+        ok ->
+            %% create a key
+            case leo_s3_auth:create_key(UserId) of
+                {ok, Keys} ->
+                    AccessKeyId = leo_misc:get_value(access_key_id, Keys),
+                    case mnesia:write(?USER_CREDENTIAL_TABLE,
+                                      #user_credential{user_id = UserId,
+                                                       access_key_id = AccessKeyId,
+                                                       created_at = CreatedAt}, write) of
+                        ok ->
+                            {ok, Keys};
+                        _ ->
+                            {error, transaction_abort}
+                    end;
+                Error ->
+                    Error
+            end;
+        _ ->
+            {error, transaction_abort}
+    end.
```

It would be good to write the erlang error handling in the `let it crash` way when writing a closure given to mnesia:activity because mensia:rollback will be called properly in case any errors (including runtime errors) happen inside the closure.

So we can rewrite like below
```erlang
ok = mnesia:write(?USERS_TABLE,
                 #?S3_USER{id = UserId,
                           password = Digest,
                           created_at = CreatedAt,
                           updated_at = CreatedAt}, write),
%% create a key
case leo_s3_auth:create_key(UserId) of
    {ok, Keys} ->
        AccessKeyId = leo_misc:get_value(access_key_id, Keys),
        ok = mnesia:write(?USER_CREDENTIAL_TABLE,
                                      #user_credential{user_id = UserId,
                                                       access_key_id = AccessKeyId,
                                                       created_at = CreatedAt}, write),
        {ok, Keys};
    Error ->
        Error
end.
```
In this way, the number of branches get smaller and make it easy to read the code.

---

I think mnesia:write called inside https://github.com/leo-project/leofs/blob/v1/apps/leo_manager/src/leo_manager_console.erl#L2119-L2121 leo_s3_user:update/1 also should be included in mnesia:activity to make the whole operation atomic.